### PR TITLE
Implement IPlatformLifetimeEvents for X11/Linux

### PR DESCRIPTION
**Pull request by jmacato**
## What does the pull request do?

Allows users to interrupt the shutdown event in Linux via X Session Management Protocol.

https://user-images.githubusercontent.com/16554748/140779523-0fbaa076-abb1-4f36-b6da-bceb4ae108c4.mp4

## What is the current behavior?

There was no way to inhibit shutdowns on Linux, unlike on win/macOS.

### CODE REVIEW

File: src/Avalonia.X11/X11PlatformLifetimeEvents.cs [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        {
+            // ReSharper disable once ConditionIsAlwaysTrueOrFalse
+            if (ShutdownRequested is null)
+            {
+                return;
+            }
+            
+            var e = new ShutdownRequestedEventArgs();
+
+            ShutdownRequested(this, e);
```
Comment (danwalmsley) [inline src/Avalonia.X11/X11PlatformLifetimeEvents.cs:229]: needs to be done on the uithead!

**Review by danwalmsley — CHANGES_REQUESTED** [review]
tested, and crashes because of invalid thread.

**Review by danwalmsley — APPROVED** [review]
tested, works LGTM

File: src/Avalonia.X11/X11PlatformLifetimeEvents.cs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+        private static readonly SMLib.SmcSaveCompleteProc s_saveCompleteDelegate = SmcSaveCompleteHandler;
+        private static readonly SMLib.SmcInteractProc s_smcInteractDelegate = StaticInteractHandler;
+        private static readonly SMLib.SmcErrorHandler s_smcErrorHandlerDelegate = StaticErrorHandler;
+        private static readonly ICELib.IceErrorHandler s_iceErrorHandlerDelegate = StaticErrorHandler;
+        private static readonly ICELib.IceIOErrorHandler s_iceIoErrorHandlerDelegate = StaticIceIOErrorHandler;
+        private static readonly SMLib.IceWatchProc s_iceWatchProcDelegate = IceWatchHandler;
+
+        private bool AllowShutdownCancellation =>
+            Environment.GetEnvironmentVariable("AVALONIA_X11_USE_SESSION_MANAGER") != "0";
```
Comment (workgroupengineering) [inline src/Avalonia.X11/X11PlatformLifetimeEvents.cs:38]: Wouldn't it be better to add a configuration in AppBuilder instead of an environment variable?

Comment (jmacato) [reply src/Avalonia.X11/X11PlatformLifetimeEvents.cs:38]: No, this is only intended to be like other X11 switches we already have. @kekekeks and @danwalmsley made this suggestion.

Comment (kekekeks) [reply src/Avalonia.X11/X11PlatformLifetimeEvents.cs:38]: @jmacato that switch should live in `X11PlatformOptions` with _default_ value determined by environment variable.

Comment (jmacato) [reply src/Avalonia.X11/X11PlatformLifetimeEvents.cs:38]: gotcha

**Review by danwalmsley — APPROVED** [review]
LGMT

**Pull request merged**